pandas/core/groupby/groupby.py

```diff
@@ -2696,6 +2697,103 @@ def _reindex_output(
 
         return output.reset_index(drop=True)
 
+    def sample(
+        self,
+        n: Optional[int] = None,
+        frac: Optional[float] = None,
+        replace: bool = False,
+        weights: Optional[Union[Sequence, Series]] = None,
+        random_state=None,
+    ):
+        """
+        Return a random sample of items from each group.
+
+        You can use `random_state` for reproducibility.
+
+        .. versionadded:: 1.1.0
+
+        Parameters
+        ----------
+        n : int, optional
+            Number of items to return. Cannot be used with `frac`.
+            Default = 1 if `frac` is None.
+        frac : float, optional
+            Fraction of items to return. Cannot be used with `n`.
+        replace : bool, default False
+            Allow or disallow sampling of the same row more than once.
+        weights : list-like, optional
+            Default None results in equal probability weighting.
```

Comment:
Do groups have to have the same number of elements?  I don't seem to be able to understand how this would work in general.

Reply:
It won't work in general, we'll get an error if n is larger than one of the group sizes and replace is False

Reply:
If I understand your example, weights has the same size as the index of the underlying DF or Series where the groupby has been applied? So in principle it could be a column from the same DF, like
```
df.groupby("a").sample(1, weights=df.weights)
```

Reply:
Maybe you should state that it must have at least 1 non-zero entry per group.

Reply:
Passing a list appears to only work if the underlying DF or Series has a standard RangeIndex(0, nobs).

Reply:
Thanks for pointing this out. Should be fixed and I updated the weights test to look at this type of case.